File: AndroidCinema/app/src/main/java/com/crpr/androidcinema/domain/discover/ListMovieModel.java
package com.crpr.androidcinema.domain.discover;

import java.io.Serializable;

/**
 * Created by claudioribeiro on 12/07/16.
 */
public class ListMovieModel implements Serializable {

    private int id;
    private String image_path;
    private double popularity;
    private String title;
    private double vote_average;

    private ListMovieModel(){}

    public int getId() {
        return id;
    }

    public String getImagePath() {
        return image_path;
    }

    public double getPopularity() {
        return popularity;
    }

    public String getTitle() {
        return title;
    }

    public double getVoteAverage() {
        return vote_average;
    }

    public interface ITitle {
        IImagePath title(String title);
    }

    public interface IImagePath {
        IPopularity imagePath(String url);
    }

    public interface IPopularity {
        IVoteAverage popularity(double popularity);
    }

    public interface IVoteAverage {
        IBuild votesAvg(double votes);
    }

    public interface IBuild {
        ListMovieModel build();
    }

    public static ITitle id(int id) {
        return new ListMovieModel.Builder(id);
    }

    private static class Builder implements ITitle, IImagePath, IPopularity, IVoteAverage, IBuild {

        private ListMovieModel instance = new ListMovieModel();

        public Builder(int id) {
            instance.id = id;
        }

        @Override
        public ListMovieModel build() {
            return instance;
        }

        @Override
        public IVoteAverage popularity(double popularity) {
            instance.popularity = popularity;
            return this;
        }

        @Override
        public IPopularity imagePath(String url) {
            instance.image_path = url;
            return this;
        }

        @Override
        public IImagePath title(String title) {
            instance.title = title;
            return this;
        }

        @Override
        public IBuild votesAvg(double votes) {
            instance.vote_average = votes;
            return this;
        }
    }
}

File: AndroidCinema/app/src/main/java/com/crpr/androidcinema/domain/discover/ListMovieModel.kt
package com.crpr.androidcinema.domain.discover

import java.io.Serializable

/**
 * Created by claudioribeiro on 12/07/16.
 */
class ListMovieModel private constructor() : Serializable {

    var id: Int = 0
        private set
    var imagePath: String? = null
        private set
    var popularity: Double = 0.toDouble()
        private set
    var title: String? = null
        private set
    var voteAverage: Double = 0.toDouble()
        private set

    interface ITitle {
        fun title(title: String): IImagePath
    }

    interface IImagePath {
        fun imagePath(url: String): IPopularity
    }

    interface IPopularity {
        fun popularity(popularity: Double): IVoteAverage
    }

    interface IVoteAverage {
        fun votesAvg(votes: Double): IBuild
    }

    interface IBuild {
        fun build(): ListMovieModel
    }

    private class Builder(id: Int) : ITitle, IImagePath, IPopularity, IVoteAverage, IBuild {

        private val instance = ListMovieModel()

        init {
            instance.id = id
        }

        override fun build(): ListMovieModel {
            return instance
        }

        override fun popularity(popularity: Double): IVoteAverage {
            instance.popularity = popularity
            return this
        }

        override fun imagePath(url: String): IPopularity {
            instance.imagePath = url
            return this
        }

        override fun title(title: String): IImagePath {
            instance.title = title
            return this
        }

        override fun votesAvg(votes: Double): IBuild {
            instance.voteAverage = votes
            return this
        }
    }

    companion object {

        fun id(id: Int): ITitle {
            return ListMovieModel.Builder(id)
        }
    }
}
